packages/backend/src/nest/tor/tor-control.service.ts

```diff
@@ -25,74 +29,96 @@ export class TorControl {
       })
+
       this.updateAuthString()
       this.connection.write(this.authString)
     })
   }
 
+  private async connect(): Promise<void> {
+    // TODO: We may want to limit the number of connection attempts.
+    while (true) {
+      try {
+        this.logger(`Connecting to Tor, host: ${this.torControlParams.host} port: ${this.torControlParams.port}`)
+        await this._connect()
+        this.logger('Tor connected')
+        return
+      } catch (e) {
+        this.logger(e)
+        this.logger('Retrying...')
+        await new Promise(r => setTimeout(r, 500))
+      }
+    }
+  }
+
   private disconnect() {
     try {
       this.connection?.end()
     } catch (e) {
-      this.logger.error('Cant disconnect', e.message)
+      this.logger.error('Disconnect failed:', e.message)
     }
     this.connection = null
   }
 
-  // eslint-disable-next-line @typescript-eslint/ban-types
-  private async _sendCommand(command: string, resolve: Function, reject: Function) {
-    await this.connect()
+  public _sendCommand(command: string): Promise<{ code: number; messages: string[] }> {
+    return new Promise((resolve, reject) => {
+      const connectionTimeout = setTimeout(() => {
+        reject('Timeout while sending command to Tor')
+      }, 5000)
+
+      this.connection?.on('data', async data => {
+        const dataArray = data.toString().split(/\r?\n/)
 
-    const connectionTimeout = setTimeout(() => {
-      reject('TOR: Send command timeout')
-    }, 5000)
-    this.connection?.on('data', async data => {
-      this.disconnect()
-      const dataArray = data.toString().split(/\r?\n/)
-      if (dataArray[0].startsWith('250')) {
-        resolve({ code: 250, messages: dataArray })
-      } else {
+        if (dataArray[0].startsWith('250')) {
+          resolve({ code: 250, messages: dataArray })
+        } else {
+          clearTimeout(connectionTimeout)
+          reject(`${dataArray[0]}`)
+        }
         clearTimeout(connectionTimeout)
-        reject(`${dataArray[0]}`)
-      }
-      clearTimeout(connectionTimeout)
+      })
+
+      this.connection?.write(command + '\r\n')
     })
-    this.connection?.write(command + '\r\n')
   }
 
   public async sendCommand(command: string): Promise<{ code: number; messages: string[] }> {
-    await this.waitForDisconnect()
-    return await new Promise((resolve, reject) => {
-      void this._sendCommand(command, resolve, reject)
-    })
-  }
+    // Only send one command at a time.
+    if (this.isSending) {
+      this.logger('Tor connection already established, waiting...')
+    }
 
-  private async waitForDisconnect() {
-    await new Promise<void>(resolve => {
-      if (!this.connection) {
-        resolve()
-      }
-    })
+    // Wait for existing command to finish.
+    while (this.isSending) {
+      await new Promise(r => setTimeout(r, 750))
+    }
+
+    this.isSending = true
+    await this.connect()
+    // FIXME: Errors are not caught here. Is this what we want?
```

Comment:
Yeah, we should handle that

Reply:
Just noting that this is how it was in the previous version of this code, so I haven't changed anything